src/io/rainrobot/maxdispachreq/AvailabilityService.java

```diff
@@ -0,0 +1,61 @@
+package io.rainrobot.maxdispachreq;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class AvailabilityService {
+
+    private int maxAvailability;
+    private Map<Integer, Set<ServerAddress>> avelabiletyServerSetMap;
+    private AddressAvelabiletyTracker addressAvelabiletyTracker;
+
+    private void recomputeMaxAvailability() {
+        if (getMaxAvailabilitySet().size() == 0) {
+            do{
+                maxAvailability --;
+            } while (getMaxAvailabilitySet().size() == 0
+                                    && maxAvailability != 1);
+            maxAvailability --;
+        }
+    }
+    
+    private Set<ServerAddress> getMaxAvailabilitySet() {
+        return avelabiletyServerSetMap.get(maxAvailability);
+    }
+
+    public ServerAddress popFromMaxAvailableSet() {
+        ServerAddress popServer = getMaxAvailabilitySet().iterator().next();
+        decrementAddressAvailability(popServer, maxAvailability);
+        recomputeMaxAvailability();
+        return popServer;
+    }
+
+    private void decrementAddressAvailability(ServerAddress popServer, int availability) {
+        avelabiletyServerSetMap.get(availability).remove(popServer);
+        avelabiletyServerSetMap.get(availability - 1).add(popServer);
+        addressAvelabiletyTracker.put(popServer, availability -1);
+    }
+
+    public boolean fullCapacity() {
+        return maxAvailability == 0;
+    }
+
+    public void initializeMap(int max_requests) {
+        for(int i = 0; i <= max_requests; i++){
```

Comment:
not clear to me what is max_request, and why you are adding hash per request num. if I put 3 then I have 3 hash sets generated.

Reply:
max_request is the maximum session a server can process (maybe should be named max_sessions).
The map hold separate HashSet value for the key of "number of session available currently for adding".

max_request = 3 
server1, server2 each hold 3 sessions each (cant take more session -> availability = 0)
server3  hold 1 session (availability = 2)
server4 dont hold any session (availability = 3)
key         val
0              {server1, server2}
1              {}
2              {server3}
3              {server4}